// @mysten/suins/dist/contracts/suins/controller.mjs
import { MoveStruct, MoveTuple, normalizeMoveArguments } from "../utils/index.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/suins/controller.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@suins/core::controller";
const ControllerV2 = new MoveTuple({
	name: `${$moduleName}::ControllerV2`,
	fields: [bcs.bool()]
});
const Controller = new MoveStruct({
	name: `${$moduleName}::Controller`,
	fields: { dummy_field: bcs.bool() }
});
/** Set the reverse lookup address for the domain */
function setReverseLookup(options) {
	const packageAddress = options.package ?? "@suins/core";
	const argumentsTypes = [null, "0x1::string::String"];
	const parameterNames = ["suins", "domainName"];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "controller",
		function: "set_reverse_lookup",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}
function burnExpired(options) {
	const packageAddress = options.package ?? "@suins/core";
	const argumentsTypes = [
		null,
		null,
		"0x2::clock::Clock"
	];
	const parameterNames = ["suins", "nft"];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "controller",
		function: "burn_expired",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}
function burnExpiredSubname(options) {
	const packageAddress = options.package ?? "@suins/core";
	const argumentsTypes = [
		null,
		null,
		"0x2::clock::Clock"
	];
	const parameterNames = ["suins", "nft"];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "controller",
		function: "burn_expired_subname",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}

//#endregion
export { burnExpired, burnExpiredSubname, setReverseLookup };
